File: app/src/main/java/ru/alexsuvorov/paistewiki/Utils/NewsLoader.java
package ru.alexsuvorov.paistewiki.Utils;

import android.os.AsyncTask;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.ArrayList;

import ru.alexsuvorov.paistewiki.model.NewsMonth;
import ru.alexsuvorov.paistewiki.model.NewsPost;

public class NewsLoader extends AsyncTask<String, Void, ArrayList<NewsMonth>> {

    public static ArrayList<NewsMonth> monthList = new ArrayList<>();
    public static ArrayList<NewsPost> postsList = new ArrayList<>();
    private String TAG = "NewsLoader";

    /*@Override
    protected void onProgressUpdate(Void values) {
        txt.setText("Running..."+ values[0]);
        progressBar.setProgress(values[0]);
    }*/

    @Override
    protected ArrayList<NewsMonth> doInBackground(String... urls) {
        String URL = urls[0];
        try {
            Document month = Jsoup.connect(URL).get();
            if (month != null) {
                //Список месяцев
                Elements monthRows = month.getElementsByClass("contlefta").select("tr");
                if (monthRows.size() > 1) {
                    for (int i = 0; i < monthRows.size(); i++) {
                        Element monthRow = monthRows.get(i);  //check all tr tags
                        Elements monthRowItems = monthRow.select("td");  //All(1), Prod, Artist
                        Element title = monthRowItems.first(); //Select All
                        Elements links = monthRowItems.select("a[href]");
                        Element link = links.first();
                        String monthUrl = "http://paiste.com/e/news.php" + link.attr("href");
                        String monthTitle = title.text();

                        //Log.d(TAG, "Link: " + monthUrl);
                        //Log.d(TAG, "Title: " + monthTitle);
                        //---------------------------------------------------
                        Document posts = Jsoup.connect(monthUrl).get();
                        if (posts != null) {
                            Elements postsRows = posts.getElementsByClass("contrighta").select("tr");
                            /*FOR IMAGES
                            http://stackoverflow.com/questions/10457415/extract-image-src-using-jsoup
                            */

                            //Если есть новости
                            if (postsRows.size() > 1) {
                                //начинать с первой новости, с 0 + пустая строка
                                for (int j = 0; j < postsRows.size(); j++) {
                                    Element row = postsRows.get(j);
                                    Elements rowItems1 = row.select("td");
                                    Elements links1 = rowItems1.select("a[href]");
                                    for (Element link1 : links1) {
                                        String linkUrl = "http://paiste.com/e/news.php" + link1.attr("href");
                                        String linkTitle = link1.text();
                                        String linkCategory = rowItems1.get(2).text();
                                        //Log.d(TAG, "Text: " + linkTitle);
                                        //Log.d(TAG, "Link: " + linkUrl);
                                        //Log.d(TAG, "Category: " + linkCategory);
                                        postsList.add(new NewsPost(linkTitle, linkCategory, linkUrl));
                                    }
                                }
                            } else {
                                //Log.d(TAG, "There is no news yet");
                                postsList.add(new NewsPost("There is no news yet", null, null));
                            }
                        }
                        monthList.add(new NewsMonth(monthTitle, monthUrl, postsList));
                        publishProgress();
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
            /*Toast toast = Toast.makeText(getContext(),
                    "Ошибка соединения!", Toast.LENGTH_SHORT);
            toast.show();*/
        }
        return monthList;

    }

    public static ArrayList<NewsMonth> getMonthList() {
        return monthList;
    }

    public static ArrayList<NewsPost> getPostsList() {
        return postsList;
    }
}

File: feature/news/impl/src/main/java/ru/alexsuvorov/paistewiki/feature/news/tools/NewsLoader.kt
package ru.alexsuvorov.paistewiki.feature.news.tools

import android.content.Context
import android.os.AsyncTask
import android.util.Log
import org.jsoup.Jsoup
import ru.alexsuvorov.paistewiki.core.database.AppDatabase
import ru.alexsuvorov.paistewiki.core.database.model.Month
import ru.alexsuvorov.paistewiki.core.database.model.News
import ru.alexsuvorov.paistewiki.core.support.AppParams
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.regex.Pattern

class NewsLoader : AsyncTask<Any?, Void, Boolean>() {

    private var monthIndex: String? = null
    private var yearIndex: String? = null
    private var context: Context? = null

    override fun doInBackground(vararg params: Any?): Boolean {
        val URL = params[0] as String
        context = params[1] as Context
        val db = AppDatabase.getDatabase(context!!)
        val newsDao = db.newsDao()
        val monthDao = db.monthDao()
        try {
            val monthDoc = Jsoup.connect(URL).get()
            if (monthDoc != null) {
                val monthRows = monthDoc.getElementsByClass("contlefta").select("tr")
                if (monthRows.size > 1) {
                    for (i in monthRows.size downTo 1) {
                        val monthRowElement = monthRows[i - 1]
                        val monthRowItems = monthRowElement.select("td")
                        val monthTitleElement = monthRowItems.first()
                        val monthLinks = monthRowItems.select("a[href]")
                        val monthLink = monthLinks.first()
                        val monthUrl = "http://paiste.com/e/news.php" + monthLink?.attr("href")
                        val monthTitle = monthTitleElement?.text() ?: ""
                        
                        val pYear = Pattern.compile("year=[0-9]{4}")
                        val mYear = pYear.matcher(monthUrl)
                        if (mYear.find()) {
                            val findY_Index1 = mYear.group(0)
                            val parts1 = findY_Index1!!.split("=")
                            yearIndex = parts1[1]
                        }
                        
                        val pMonth = Pattern.compile("month=\\d{1,12}")
                        val mMonth = pMonth.matcher(monthUrl)
                        if (mMonth.find()) {
                            val findM_Index1 = mMonth.group(0)
                            val clearedValue = clearValues(findM_Index1!!)
                            monthIndex = if (clearedValue.length == 1) {
                                "0$clearedValue"
                            } else {
                                clearedValue
                            }
                        }
                        
                        val mIndex = (yearIndex!! + monthIndex!!).toInt()
                        Log.d(javaClass.simpleName, "INDEX: $mIndex")
                        
                        val postsDoc = Jsoup.connect(monthUrl).get()
                        if (postsDoc != null) {
                            val postsRows = postsDoc.getElementsByClass("contrighta").select("tr")
                            if (postsRows.size > 1) {
                                for (j in 0 until postsRows.size) {
                                    val postsRowElement = postsRows[j]
                                    val postsRowItems = postsRowElement.select("td")
                                    val postsLinks = postsRowItems.select("a[href]")
                                    for (postLink in postsLinks) {
                                        val linkUrl = "http://paiste.com/e/news.php" + postLink.attr("href")
                                        val linkTitle = postLink.text()
                                        val linkCategory = if (postsRowItems.size > 2) postsRowItems[2].text() else ""
                                        if (newsDao.insert(News(0, linkTitle, linkCategory, linkUrl, mIndex.toLong())) > 0) {
                                            AppParams.newsUpdated = true
                                        }
                                    }
                                }
                            }
                            monthDao.insert(Month(monthDao.getLastMonthId().toLong() + 1, monthTitle, monthUrl, mIndex))
                        }
                    }
                }
            }
        } catch (exception: SocketTimeoutException) {
            exception.printStackTrace()
            AppParams.errorCode = 1
            return true
        } catch (exception: UnknownHostException) {
            exception.printStackTrace()
            AppParams.errorCode = 2
            Log.d(javaClass.simpleName, "Server error")
            return true
        } catch (exception: IOException) {
            exception.printStackTrace()
            AppParams.errorCode = 3
            return true
        }
        AppParams.errorCode = 0
        return true
    }

    private fun clearValues(input: String): String {
        val parts = input.split("=")
        return parts[1]
    }

    override fun onPostExecute(result: Boolean) {
        super.onPostExecute(result)
        Log.d(javaClass.simpleName, "Result is: $result")
    }
}